### DISCUSSION TITLE
CMake + pch + ccache leads to unnecessary cache misses?

### DISCUSSION THREAD

**Discussion by KaiSalm** [ccache/ccache#1188] [Support]
When using the target_precompile_headers feature of cmake, it produces the precompiled header with a call like

```
ccache /usr/bin/c++ \
-Winvalid-pch -x c++-header \
-include CMakeFiles/foo.dir/cmake_pch.hxx \
-MD -MT CMakeFiles/foo.dir/cmake_pch.hxx.gch \
-MF CMakeFiles/foo.dir/cmake_pch.hxx.gch.d \
-o CMakeFiles/foo.dir/cmake_pch.hxx.gch \
-c CMakeFiles/foo.dir/cmake_pch.hxx.cxx
```

which is already a bit weird since it does not just compile the header, but it works fine, even with ccache everything is fine on the initial build.

If i however change anything that causes the pch to be rebuild, on the second parsing of the command line there exists already a .gch file. This means that ccache assumes that the -include causes the compiler to read the gch instead. gcc however is as far as i can tell clever enough to notice that it might be a bad idea to use a .gch that will be overwritten by the very call it is currently executing, so that it does not use the gch. With ccache assuming that it does, ccache includes the hash of the previous gch into the hashed data, which causes cache misses down the line if you want to rebuild the pch.

Have i understood this correctly? Would you consider this a bug in ccache, or is your opinion that the commandline produced by CMake is the root cause?

I'm not sure what the best fix would be. Locally i tested checking whether the output file is the assumed pch file and if so disregarding it, which seems to work, but obviously i do not know whether there will be unintended consequences in a system so big...

**Comment by jrosdahl** [comment]
> Have i understood this correctly?

The explanation sounds reasonable to me.

> Would you consider this a bug in ccache, or is your opinion that the commandline produced by CMake is the root cause?

I wouldn't say that it's a bug in ccache but a missing feature.

**Reply by KaiSalm** [reply]
Then consider this a feature request? ;) Or should i open a seperate request?

**Reply by jrosdahl** [reply]
I'll create a separate feature request.

I had a quick look at this using `strace(1)`, and it seems like GCC and Clang behave differently for `$compiler -include X.h -c X.h`:

* GCC never reads `X.h.gch`, not even when compiling another header with `gcc -include X.h -c Y.h`.
* Clang always reads `X.h.gch` (if it exists), both when compiling `X.h` itself and when compiling another header.

That leads to the question what CMake generates for Clang since a command similar to that of GCC can't work, and yes, it generates something like this instead: `clang -Xclang -emit-pch -Xclang -include -Xclang X.h -c X.h`. This incantation makes Clang not read `X.h.gch`. :roll_eyes:

**Reply by jrosdahl** [reply]
> I'll create a separate feature request.

#1196.
